From 87d059a12ff77f4564cd08febe64a8c80e7b21c4 Mon Sep 17 00:00:00 2001
From: "Herman S." <429230+has207@users.noreply.github.com>
Date: Mon, 27 Jul 2026 15:12:11 +0900
Subject: [PATCH 4/5] [CPU] Emit a real acquire barrier for isync and drop
 lwarx/stwcx fences

PS5X360E: ported from Xenia Edge (github.com/has207/xenia-edge) onto
Xenia Canary b083312b + PS5X360's patch. Unchanged otherwise.

(cherry picked from commit 9c729844c02c5d8c90feb901c644eab289a19885)
---
 src/xenia/cpu/backend/a64/a64_seq_memory.cc | 10 +++++++++
 src/xenia/cpu/backend/x64/x64_seq_memory.cc |  9 ++++++++
 src/xenia/cpu/hir/hir_builder.cc            |  2 ++
 src/xenia/cpu/hir/hir_builder.h             |  1 +
 src/xenia/cpu/hir/opcodes.h                 |  1 +
 src/xenia/cpu/hir/opcodes.inl               |  8 +++++++
 src/xenia/cpu/ppc/ppc_emit_memory.cc        | 24 +++------------------
 7 files changed, 34 insertions(+), 21 deletions(-)

diff --git a/src/xenia/cpu/backend/a64/a64_seq_memory.cc b/src/xenia/cpu/backend/a64/a64_seq_memory.cc
index 47d04a71f..0560ae41e 100644
--- a/src/xenia/cpu/backend/a64/a64_seq_memory.cc
+++ b/src/xenia/cpu/backend/a64/a64_seq_memory.cc
@@ -69,6 +69,16 @@ struct MEMORY_BARRIER
 };
 EMITTER_OPCODE_TABLE(OPCODE_MEMORY_BARRIER, MEMORY_BARRIER);
 
+// ============================================================================
+// OPCODE_LOAD_BARRIER
+// ============================================================================
+struct LOAD_BARRIER : Sequence<LOAD_BARRIER, I<OPCODE_LOAD_BARRIER, VoidOp>> {
+  static void Emit(A64Emitter& e, const EmitArgType& i) {
+    e.dmb(Xbyak_aarch64::ISHLD);
+  }
+};
+EMITTER_OPCODE_TABLE(OPCODE_LOAD_BARRIER, LOAD_BARRIER);
+
 // ============================================================================
 // OPCODE_CACHE_CONTROL
 // ============================================================================
diff --git a/src/xenia/cpu/backend/x64/x64_seq_memory.cc b/src/xenia/cpu/backend/x64/x64_seq_memory.cc
index 091b5bb9b..e8977070e 100644
--- a/src/xenia/cpu/backend/x64/x64_seq_memory.cc
+++ b/src/xenia/cpu/backend/x64/x64_seq_memory.cc
@@ -2038,6 +2038,15 @@ struct MEMORY_BARRIER
 };
 EMITTER_OPCODE_TABLE(OPCODE_MEMORY_BARRIER, MEMORY_BARRIER);
 
+// ============================================================================
+// OPCODE_LOAD_BARRIER
+// ============================================================================
+struct LOAD_BARRIER : Sequence<LOAD_BARRIER, I<OPCODE_LOAD_BARRIER, VoidOp>> {
+  // x86 never reorders a load with a later access, so nothing to emit.
+  static void Emit(X64Emitter& e, const EmitArgType& i) {}
+};
+EMITTER_OPCODE_TABLE(OPCODE_LOAD_BARRIER, LOAD_BARRIER);
+
 // ============================================================================
 // OPCODE_MEMSET
 // ============================================================================
diff --git a/src/xenia/cpu/hir/hir_builder.cc b/src/xenia/cpu/hir/hir_builder.cc
index 26bd58195..0031d30a0 100644
--- a/src/xenia/cpu/hir/hir_builder.cc
+++ b/src/xenia/cpu/hir/hir_builder.cc
@@ -1355,6 +1355,8 @@ void HIRBuilder::CacheControl(Value* address, size_t cache_line_size,
 
 void HIRBuilder::MemoryBarrier() { AppendInstr(OPCODE_MEMORY_BARRIER_info, 0); }
 
+void HIRBuilder::LoadBarrier() { AppendInstr(OPCODE_LOAD_BARRIER_info, 0); }
+
 void HIRBuilder::DelayExecution() {
   AppendInstr(OPCODE_DELAY_EXECUTION_info, 0);
 }
diff --git a/src/xenia/cpu/hir/hir_builder.h b/src/xenia/cpu/hir/hir_builder.h
index e8a80415a..ca76ffe7a 100644
--- a/src/xenia/cpu/hir/hir_builder.h
+++ b/src/xenia/cpu/hir/hir_builder.h
@@ -208,6 +208,7 @@ class HIRBuilder {
   void CacheControl(Value* address, size_t cache_line_size,
                     CacheControlType type);
   void MemoryBarrier();
+  void LoadBarrier();
   void DelayExecution();
   void SetRoundingMode(Value* value);
   Value* Max(Value* value1, Value* value2);
diff --git a/src/xenia/cpu/hir/opcodes.h b/src/xenia/cpu/hir/opcodes.h
index 43be8810a..7500bf070 100644
--- a/src/xenia/cpu/hir/opcodes.h
+++ b/src/xenia/cpu/hir/opcodes.h
@@ -292,6 +292,7 @@ enum Opcode {
   OPCODE_DELAY_EXECUTION,  // for db16cyc
   OPCODE_RESERVED_LOAD,
   OPCODE_RESERVED_STORE,
+  OPCODE_LOAD_BARRIER,
 
   __OPCODE_MAX_VALUE,  // Keep at end.
 };
diff --git a/src/xenia/cpu/hir/opcodes.inl b/src/xenia/cpu/hir/opcodes.inl
index e0c31e577..53b618474 100644
--- a/src/xenia/cpu/hir/opcodes.inl
+++ b/src/xenia/cpu/hir/opcodes.inl
@@ -278,6 +278,14 @@ DEFINE_OPCODE(
     OPCODE_SIG_X,
     OPCODE_FLAG_MEMORY | OPCODE_FLAG_VOLATILE)
 
+// Orders prior loads against later accesses, the acquire half of a lock.
+// Free on a store-ordered host.
+DEFINE_OPCODE(
+    OPCODE_LOAD_BARRIER,
+    "load_barrier",
+    OPCODE_SIG_X,
+    OPCODE_FLAG_MEMORY | OPCODE_FLAG_VOLATILE)
+
 DEFINE_OPCODE(
     OPCODE_MAX,
     "max",
diff --git a/src/xenia/cpu/ppc/ppc_emit_memory.cc b/src/xenia/cpu/ppc/ppc_emit_memory.cc
index 8483ce1d5..f4de593f6 100644
--- a/src/xenia/cpu/ppc/ppc_emit_memory.cc
+++ b/src/xenia/cpu/ppc/ppc_emit_memory.cc
@@ -757,8 +757,9 @@ int InstrEmit_sync(PPCHIRBuilder& f, const InstrData& i) {
 }
 
 int InstrEmit_isync(PPCHIRBuilder& f, const InstrData& i) {
-  // XEINSTRNOTIMPLEMENTED();
-  f.Nop();
+  // Only the memory ordering half is modelled. Guests pair isync with a
+  // conditional branch to acquire a lock.
+  f.LoadBarrier();
   return 0;
 }
 
@@ -773,15 +774,12 @@ int InstrEmit_ldarx(PPCHIRBuilder& f, const InstrData& i) {
   // RESERVE_ADDR <- real_addr(EA)
   // RT <- MEM(EA, 8)
 
-  // The barrier is here so we read a current value.
   Value* ea = CalculateEA_0(f, i.X.RA, i.X.RB);
 
   if (cvars::no_reserved_ops) {
     f.StoreGPR(i.X.RT, f.ByteSwap(f.Load(ea, INT64_TYPE)));
 
   } else {
-    f.MemoryBarrier();
-
     Value* rt = f.ByteSwap(f.LoadWithReserve(ea, INT64_TYPE));
     f.StoreGPR(i.X.RT, rt);
   }
@@ -799,16 +797,12 @@ int InstrEmit_lwarx(PPCHIRBuilder& f, const InstrData& i) {
   // RESERVE_ADDR <- real_addr(EA)
   // RT <- i32.0 || MEM(EA, 4)
 
-  // The barrier is here so we read a current value.
-
   Value* ea = CalculateEA_0(f, i.X.RA, i.X.RB);
   if (cvars::no_reserved_ops) {
     f.StoreGPR(i.X.RT,
                f.ZeroExtend(f.ByteSwap(f.Load(ea, INT32_TYPE)), INT64_TYPE));
 
   } else {
-    f.MemoryBarrier();
-
     Value* rt =
         f.ZeroExtend(f.ByteSwap(f.LoadWithReserve(ea, INT32_TYPE)), INT64_TYPE);
     f.StoreGPR(i.X.RT, rt);
@@ -843,12 +837,6 @@ int InstrEmit_stdcx(PPCHIRBuilder& f, const InstrData& i) {
   }
   f.StoreContext(offsetof(PPCContext, cr0.cr0_lt), f.LoadZeroInt8());
   f.StoreContext(offsetof(PPCContext, cr0.cr0_gt), f.LoadZeroInt8());
-
-  // Issue memory barrier for when we go out of lock and want others to see our
-  // updates.
-  if (!cvars::no_reserved_ops) {
-    f.MemoryBarrier();
-  }
   return 0;
 }
 
@@ -881,12 +869,6 @@ int InstrEmit_stwcx(PPCHIRBuilder& f, const InstrData& i) {
   f.StoreContext(offsetof(PPCContext, cr0.cr0_lt), f.LoadZeroInt8());
   f.StoreContext(offsetof(PPCContext, cr0.cr0_gt), f.LoadZeroInt8());
 
-  // Issue memory barrier for when we go out of lock and want others to see our
-  // updates.
-  if (!cvars::no_reserved_ops) {
-    f.MemoryBarrier();
-  }
-
   return 0;
 }
 // Floating-point load (A-19)
